src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs

```diff
@@ -24,33 +65,82 @@ private static long ParseContentLength(string value)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private void SetValueUnknown(string key, in StringValues value)
+        private void AppendContentLength(Span<byte> value)
         {
-            Unknown[key] = value;
+            if (_contentLength.HasValue)
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.MultipleContentLengths);
+            }
+
+
+            if (!TryParseNonNegativeInt64(value, out var parsed))
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.InvalidContentLength, value.GetAsciiOrUTF8StringNonNullCharacters());
+            }
+
+            _contentLength = parsed;
         }
 
-        public unsafe void Append(Span<byte> name, string value)
+        private static unsafe bool TryParseNonNegativeInt64(Span<byte> value, out long result)
         {
-            fixed (byte* namePtr = name)
+            if (value.Length == 0)
+            {
+                result = 0;
+                return false;
+            }
+
+            var calc = 0L;
+            var i = 0;
+            for (; i < value.Length; i++)
+            {
+                var digit = (long)value[i] - (long)0x30;
+                if ((ulong)digit > 9)
+                {
+                    // Not a digit
+                    break;
+                }
+
+                calc = calc * 10 + digit;
+                if (calc < 0)
+                {
+                    // Overflow
+                    break;
+                }
+            }
+
+            if (i != value.Length)
             {
-                Append(namePtr, name.Length, value);
+                // Didn't parse correct until end
+                result = 0;
+                return false;
             }
+
+            result = calc;
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void SetValueUnknown(string key, in StringValues value)
+        {
+            Unknown[key] = value;
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private unsafe void AppendUnknownHeaders(byte* pKeyBytes, int keyLength, string value)
+        private unsafe void AppendUnknownHeaders(Span<byte> name, Span<byte> value)
         {
-            string key = new string('\0', keyLength);
+            string key = new string('\0', name.Length);
```

Comment:
string.Create?

Reply:
Complicated as can't pass the Span as the `TState` so would still need to fix it to pass as `byte*`; then would need to fix the `Span<char>` in the `SpanAction` pass to `TryGetAsciiString` as it currently is; so would be a bigger change to address all that.

Reply:
https://github.com/aspnet/AspNetCore/issues/8978